NIFI-10460 GetZendesk processor

briansolo1985 opened the pull request:
<!-- Licensed to the Apache Software Foundation (ASF) under one or more -->
<!-- contributor license agreements.  See the NOTICE file distributed with -->
<!-- this work for additional information regarding copyright ownership. -->
<!-- The ASF licenses this file to You under the Apache License, Version 2.0 -->
<!-- (the "License"); you may not use this file except in compliance with -->
<!-- the License.  You may obtain a copy of the License at -->
<!--     http://www.apache.org/licenses/LICENSE-2.0 -->
<!-- Unless required by applicable law or agreed to in writing, software -->
<!-- distributed under the License is distributed on an "AS IS" BASIS, -->
<!-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. -->
<!-- See the License for the specific language governing permissions and -->
<!-- limitations under the License. -->

# Summary
Adding a processor which can fetch data from Zendesk Incremental Export API.

[NIFI-10460](https://issues.apache.org/jira/browse/NIFI-10460)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x]  Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [ ] JDK 11
  - [ ] JDK 17

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

arpadboda commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/main/java/org/apache/nifi/processors/zendesk/GetZendesk.java:
```diff
@@ -0,0 +1,375 @@
+            }
+        } else if (response.statusCode() == HTTP_TOO_MANY_REQUESTS) {
+            context.yield();
+            throw new ProcessException(format("Rate limit exceeded, yielding before retrying request."));
```
What's the goal of throwing here?
ProcessSession is going to catch this exception, roll back the session and make the processor yield anyway. But there is no flowfile related operation in this scenario to roll back, so other than logging, nothing is going to happen here.

briansolo1985 replied:
Thanks for the highlight. There was no particular intention with the throw. Will remove the throw clause, and log the error instead.

arpadboda reviewed (commented):
Mostly LGTM, added some minor comments

briansolo1985 commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/main/java/org/apache/nifi/processors/zendesk/GetZendesk.java:
```diff
@@ -0,0 +1,375 @@
+
+        String password = validationContext.getProperty(ZENDESK_PASSWORD).evaluateAttributeExpressions().getValue();
+        String token = validationContext.getProperty(ZENDESK_TOKEN).evaluateAttributeExpressions().getValue();
+        if (isNotBlank(password) && isNotBlank(token)) {
```
This is needed to handle the case when both parameters are set, which is illegal. I didn't find any options to handle dependency with other property in property level validator (maybe I missed something).
Updated the explanation to make it more expressive.

briansolo1985 commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/main/java/org/apache/nifi/processors/zendesk/GetZendesk.java:
```diff
@@ -0,0 +1,375 @@
+    private static final PropertyDescriptor ZENDESK_PASSWORD = new PropertyDescriptor.Builder()
+        .name(ZENDESK_PASSWORD_NAME)
+        .displayName("Zendesk Password")
+        .description("Password of Zendesk login user. Leave empty if token is provided.")
```
Thanks for spotting this. I will add '.' at the and of all descriptions

briansolo1985 commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/main/java/org/apache/nifi/processors/zendesk/GetZendesk.java:
```diff
@@ -0,0 +1,375 @@
+            context.yield();
+            throw new ProcessException(format("Rate limit exceeded, yielding before retrying request."));
+        } else {
+            getLogger().error("HTTP {} error for uri={} with response={}", response.statusCode(), uri, responseBodyToString(context, response));
```
Good point. Will add a yield here

briansolo1985 commented:
@arpadboda Thanks for your commits. I incorporated them in my last commit. I had to force push because I realised I made a mistake in the commits message (Jira ticket number was wrong). I also updated the PR title according to this. Sorry for any inconveniences caused.

bejancsaba commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/test/java/org/apache/nifi/processors/zendesk/GetZendeskTest.java:
```diff
@@ -0,0 +1,276 @@
+        testRunner.enableControllerService(standardWebClientServiceProvider);
+
+        testRunner.setProperty(WEB_CLIENT_SERVICE_PROVIDER_NAME, standardWebClientServiceProviderId);
+        testRunner.setProperty(ZENDESK_SUBDOMAIN_NAME, "zendesk-subdomain");
```
Is it possible to use the constants defined in the main class?

briansolo1985 replied:
Maybe I'm missing something, but the property names constants are used from the main class. The property values are arbitrary string, not relevant for the result of the tests, need to be set only because those are required properties.

bejancsaba replied:
My bad, just the name was the same as the constant and it threw me off. It is good as is.

bejancsaba commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/test/java/org/apache/nifi/processors/zendesk/GetZendeskTest.java:
```diff
@@ -0,0 +1,276 @@
+        testRunner.setProperty(ZENDESK_EXPORT_METHOD_NAME, CURSOR.getValue());
+
+        // when + then
+        AssertionError wrapperError = assertThrows(AssertionError.class, () -> testRunner.run(1));
```
I tried to execute the tests locally but this seems to be failing (all others pass) could you please take a look?

briansolo1985 replied:
Good catch, I changed the code in the previous commit, and forgot to update the tests.

bejancsaba replied:
Thanks for the update.

bejancsaba reviewed (changes requested):
Thanks for the new addition, looks really good. I noted a few things and in addition I wanted to ask whether this could be added to the assembly as well.

briansolo1985 commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/pom.xml:
```diff
@@ -0,0 +1,93 @@
+        <dependency>
+            <groupId>org.apache.nifi</groupId>
+            <artifactId>nifi-api</artifactId>
+            <version>1.18.0-SNAPSHOT</version>
```
I removed all explicit version I could - some had to be left in place, otherwise Maven throws an error

briansolo1985 commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/main/resources/org.apache.nifi.processors.zendesk.GetZendesk/additionalDetails.html:
```diff
@@ -0,0 +1,71 @@
+<h3>Export methods</h3>
+
+<p>
+    Zendesk Incremental Export API supports cursor and time based export methods.
```
Fixed

briansolo1985 commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/main/java/org/apache/nifi/processors/zendesk/GetZendesk.java:
```diff
@@ -0,0 +1,376 @@
+import org.apache.nifi.web.client.api.HttpUriBuilder;
+import org.apache.nifi.web.client.provider.api.WebClientServiceProvider;
+
+@PrimaryNodeOnly
```
Yes there is a global state for the processor holding the cursor. Also Zendesk API has strict rate limiting (eg a few requests per minutes, so concurrency here wouldn't be beneficial at all)

bejancsaba replied:
Thanks for the context

briansolo1985 commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/main/java/org/apache/nifi/processors/zendesk/GetZendesk.java:
```diff
@@ -0,0 +1,376 @@
+        .addValidator(NON_EMPTY_VALIDATOR)
+        .build();
+
+    private static final PropertyDescriptor ZENDESK_TOKEN = new PropertyDescriptor.Builder()
```
Changed this to authentication type and authentication credential. It seems to be more user friendly and resulted in less code

briansolo1985 commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/main/java/org/apache/nifi/processors/zendesk/GetZendesk.java:
```diff
@@ -0,0 +1,376 @@
+    ));
+
+    private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+    private static final JsonFactory JSON_FACTORY = OBJECT_MAPPER.getFactory();
```
The intention was to keep the processor as simple in the first iteration. Zendesk API returns an array of JSON items which then can be turned into single records by utilising SplitJSON processor. Then the data can be converted by ConvertJSONToAvro, or processed by json manipulating processor.
As Nifi provided this functionality out of the box I didn't want to put this logic into the processor.
In upcoming iterations we can extend the processor capability if there will be user interest.

bejancsaba replied:
ok, makes sense

briansolo1985 commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/main/java/org/apache/nifi/processors/zendesk/GetZendesk.java:
```diff
@@ -0,0 +1,376 @@
+            getLogger().error("Rate limit exceeded for uri={}, yielding before retrying request.", uri);
+            context.yield();
+        } else {
+            getLogger().error("HTTP {} error for uri={} with response={}.", response.statusCode(), uri, responseBodyToString(context, response));
```
Added

briansolo1985 commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/main/java/org/apache/nifi/processors/zendesk/GetZendesk.java:
```diff
@@ -0,0 +1,376 @@
+        try {
+            return context.getStateManager().getState(CLUSTER).get(zendeskResource.getValue() + exportMethod.getValue());
+        } catch (IOException e) {
+            throw new ProcessException("Failed to retrieve cursor state", e);
```
I think yield happens automatically when ProcessException is thrown (based on this comment https://github.com/apache/nifi/pull/6411#discussion_r976366765)

bejancsaba replied:
I see, thank you

briansolo1985 commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/main/java/org/apache/nifi/processors/zendesk/GetZendesk.java:
```diff
@@ -0,0 +1,376 @@
+        };
+    }
+
+    private int extractZendeskResourceData(JsonParser parser, JsonGenerator generator) throws IOException {
```
Please check https://github.com/apache/nifi/pull/6411#discussion_r977989576

briansolo1985 commented on nifi-nar-bundles/nifi-zendesk-bundle/nifi-zendesk-processors/src/main/java/org/apache/nifi/processors/zendesk/ZendeskExportMethod.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    public static ZendeskExportMethod forName(String methodName) {
+        return Stream.of(values()).filter(m -> m.getValue().equals(methodName)).findFirst()
```
Fixed

arpadboda reviewed (approved):
I'm good and happy to merge as soon as @bejancsaba is happy with this, too.

bejancsaba commented:
> I'm good and happy to merge as soon as @bejancsaba is happy with this, too.

@arpadboda this PR is good from my side. I'm not sure whether anyone with more NiFi expertise will take a look or not.

bejancsaba reviewed (approved):
+1 from my side. Thank you for applying the changes.
